server/src/main/java/org/elasticsearch/health/metadata/HealthMetadataService.java

```diff
@@ -102,61 +109,49 @@ private void registerListeners() {
                     setting,
-                    value -> updateOnSettingsUpdated(new UpdateShardLimitsMetadata(setting.getKey(), value))
+                    value -> updateOnShardLimitsSettingsUpdated(setting.getKey(), value)
                 )
             );
-
-        clusterService.getClusterSettings().addSettingsUpdateConsumer(ENABLED_SETTING, this::enable);
     }
 
-    private void enable(boolean enabled) {
+    private void updateOnHealthNodeEnabledChange(boolean enabled) {
         this.enabled = enabled;
         if (this.enabled) {
             clusterService.addListener(clusterStateListener);
-            resetHealthMetadata("health-node-enabled");
+
+            if (canPostClusterStateUpdates(clusterService.state())) {
+                taskQueue.submitTask("health-node-enabled", () -> this.localHealthMetadata, null);
+            }
         } else {
             clusterService.removeListener(clusterStateListener);
-            readyToPublish = false;
         }
     }
 
-    private void updateOnClusterStateChange(ClusterChangedEvent event) {
-        final boolean wasMaster = event.previousState().nodes().isLocalNodeElectedMaster();
-        isMaster = event.localNodeMaster();
-        if (isMaster && wasMaster == false) {
-            readyToPublish = true;
-        } else if (isMaster == false) {
-            readyToPublish = false;
-        }
+    private boolean canPostClusterStateUpdates(ClusterState state) {
         // Wait until every node in the cluster is upgraded to 8.5.0 or later
-        if (event.state().nodesIfRecovered().getMinNodeVersion().onOrAfter(Version.V_8_5_0)) {
-            if (readyToPublish) {
-                resetHealthMetadata("health-metadata-update-master-election");
-                readyToPublish = false;
-            }
-        }
+        return isMaster && state.nodesIfRecovered().getMinNodeVersion().onOrAfter(Version.V_8_5_0);
     }
 
-    private void updateOnSettingsUpdated(UpsertHealthMetadataTask task) {
-        // We do not use the cluster state to check if this is the master node because the cluster state might not have been initialized
-        if (isMaster && enabled) {
-            ClusterState clusterState = clusterService.state();
-            if (clusterState.nodesIfRecovered().getMinNodeVersion().onOrAfter(Version.V_8_5_0)) {
-                taskQueue.submitTask("health-metadata-update", task, null);
-            }
+    private void updateOnClusterStateChange(ClusterChangedEvent event) {
+        // Avoid scheduling updates to the taskQueue in case the cluster is recovering.
+        if (event.state().blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK)) {
+            return;
         }
-    }
 
-    private void resetHealthMetadata(String source) {
-        taskQueue.submitTask(source, new InsertHealthMetadata(clusterService.getClusterSettings()), null);
+        isMaster = event.localNodeMaster();
```

Comment:
This will always perform a volatile write. 

Shall we avoid it ? (note that volatile reads are much cheaper - equivalent to a memory load - than a volatile write )

```suggestion
     final boolean prevIsMaster = this.isMaster;
     if (prevIsMaster != event.localNodeMaster()) {
        this.isMaster = event.localNodeMaster(); 
      }
```